tests/unit/autocomplete/test_indexer.py

```diff
@@ -0,0 +1,180 @@
+    def setUp(self):
+        self.session = mock.Mock(spec=Session)
+        self.aws_command = DummyCommand(
+            # The CLIDriver doesn't actually have a ``name`` property,
+            # but it should probably be 'aws'.
+            command_name=None,
+            arg_table={
+                'region': DummyArg('region'),
+                'endpoint-url': DummyArg('endpoint-url'),
+            },
+            subcommand_table={
+                'ec2': DummyCommand(
+                    command_name='ec2',
+                    subcommand_table={
+                        'describe-instances': DummyCommand(
+                            'describe-instances',
+                            arg_table={
+                                'instance-ids': DummyArg('instance-ids'),
+                                'filters': DummyArg(
+                                    'filters', 'list', nargs='+'),
+                            }
+                        ),
+                        'run-instances': DummyCommand('run-instances'),
+                    }
+                ),
+                's3': DummyCommand(
+                    command_name='s3',
+                    subcommand_table={
+                        'list-objects': DummyCommand('list-objects'),
+                        'put-object': DummyCommand('put-object'),
+                    }
+                )
+            }
+        )
+        # Ideally we just use ':memory:', but that's for
+        # a specific sqlite3 connection so we'd have to
+        # update our interfaces to return/accept a db connection.
+        # I'd prefer not to have db connections as part of the
+        # indexing interface.
+        self.tempfile = tempfile.NamedTemporaryFile('w')
+        self.db_conn = indexer.DatabaseConnection(self.tempfile.name)
+        self.indexer = indexer.ModelIndexer(self.db_conn)
+        self.query = model.ModelIndex(self.tempfile.name)
+
+    def tearDown(self):
+        self.tempfile.close()
+
+    def test_can_retrieve_top_level_commands(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.command_names(lineage=['aws'])),
+            set(['ec2', 's3'])
+        )
+
+    def test_can_retrieve_operation_names(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.command_names(lineage=['aws', 'ec2'])),
+            set(['describe-instances', 'run-instances']),
+        )
+
+    def test_can_retrieve_global_params(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.arg_names(lineage=[], command_name='aws')),
+            set(['region', 'endpoint-url']),
+        )
+
+    def test_can_retrieve_service_params(self):
+        self.indexer.generate_index(self.aws_command)
+        self.assertEqual(
+            set(self.query.arg_names(lineage=['aws.ec2'],
```

Comment:
Is this a valid way to use `lineage`? It will work because we are joining on `.` under the hood, but if we change it, there will be problems and I'm not sure we want the tests to reflect this.

Reply:
No it's not, good catch.  I'll update this.